compiler/ccgexprs.nim

```diff
@@ -1938,21 +1946,33 @@ proc genCast(p: BProc, e: PNode, d: var TLoc) =
     # C code; plus it's the right thing to do for closures:
     genSomeCast(p, e, d)
 
-proc genRangeChck(p: BProc, n: PNode, d: var TLoc, magic: string) =
+proc genRangeChck(p: BProc, n: PNode, d: var TLoc) =
   var a: TLoc
   var dest = skipTypes(n.typ, abstractVar)
+  initLocExpr(p, n[0], a)
   if optRangeCheck notin p.options or (dest.kind in {tyUInt..tyUInt64} and
       checkUnsignedConversions notin p.config.legacyFeatures):
-    initLocExpr(p, n[0], a)
-    putIntoDest(p, d, n, "(($1) ($2))" %
-        [getTypeDesc(p.module, dest), rdCharLoc(a)], a.storage)
+    discard "no need to generate a check because it was disabled"
   else:
-    let mm = if dest.kind in {tyUInt32, tyUInt64, tyUInt}: "chckRangeU" else: magic
-    initLocExpr(p, n[0], a)
-    putIntoDest(p, d, lodeTyp dest, ropecg(p.module, "(($1)#$5($2, $3, $4))", [
-        getTypeDesc(p.module, dest), rdCharLoc(a),
-        genLiteral(p, n[1], dest), genLiteral(p, n[2], dest),
-        mm]), a.storage)
+    let raiser =
+      case skipTypes(n.typ, abstractVarRange).kind
+      of tyUInt..tyUInt64, tyChar: "raiseRangeErrorU"
+      of tyFloat..tyFloat128: "raiseRangeErrorF"
+      else: "raiseRangeErrorI"
+    discard cgsym(p.module, raiser)
+    # This seems to be bug-compatible with Nim version 1 but what we
+    # should really do here is to check if uint64Value < high(int)
+    let boundaryCast =
+      if n[0].typ.skipTypes(abstractVarRange).kind in {tyUInt, tyUInt32, tyUInt64}:
+        "(NI64)"
+      else:
+        ""
+    # emit range check:
+    linefmt(p, cpsStmts, "if ($6($1) < $2 || $6($1) > $3){ $4($1, $2, $3); $5}$n",
```

Comment:
seems fishy.

```nim
  proc main() =
    var z = 10.uint32
    var c = (z * (z * z + 2)).int32
  main()
```
after PR:
```
N_LIB_PRIVATE N_NIMCALL(void, main__WQ7U0WVR9aE6i9cQgC4ymCIA)(void) {
	NU32 z;
	NI32 c;
	z = ((NU32) 10);
	/*FOO*/  if ((NI64)((NU32)((NU32)(z) * (NU32)((NU32)((NU32)((NU32)((NU32)(z) * (NU32)(z))) + (NU32)(((NU32) 2)))))) < ((NI32) (-2147483647 -1)) || (NI64)((NU32)((NU32)(z) * (NU32)((NU32)((NU32)((NU32)((NU32)(z) * (NU32)(z))) + (NU32)(((NU32) 2)))))) > ((NI32) 2147483647)){ raiseRangeErrorI((NU32)((NU32)(z) * (NU32)((NU32)((NU32)((NU32)((NU32)(z) * (NU32)(z))) + (NU32)(((NU32) 2))))), ((NI32) (-2147483647 -1)), ((NI32) 2147483647)); }
	c = ((NI32) ((NU32)((NU32)(z) * (NU32)((NU32)((NU32)((NU32)((NU32)(z) * (NU32)(z))) + (NU32)(((NU32) 2)))))));
}

```

before PR:
```nim
N_LIB_PRIVATE N_NIMCALL(void, main__WQ7U0WVR9aE6i9cQgC4ymCIA)(void) {
	NU32 z;
	NI32 c;
	z = ((NU32) 10);
	c = ((NI32)chckRange((NU32)((NU32)(z) * (NU32)((NU32)((NU32)((NU32)((NU32)(z) * (NU32)(z))) + (NU32)(((NU32) 2))))), ((NI32) (-2147483647 -1)), ((NI32) 2147483647)));
}
```

Reply:
This is not a new problem though and we store the results of every real function evaluation in a temporary.

Reply:
it is a new problem here, before PR it would evaluate just once (bc function call)
in other cases maybe it was existing; anyhow this can be addressed in subsequent PR's
also just realized `chckRange` + friends are not {.inline.}, but should...

Reply:
This isn't a problem in practice. We store the results in temporaries anyways

Reply:
what do you mean? 
the cgen'd C code is now:
```nim
if (expr(z) < a || expr(z) > b) raiseRangeErrorI(expr(z), a, b)
c = expr(z)
```
I don't what temporary you're referring to.
The cgen'd C code should instead be:
```nim
c = expr(z)
if (c < a || c > b) raiseRangeErrorI(c, a, b)
```